// openai/src/shims/node.ts

import * as types from '../_shims/node-types';
import { setShims } from '../_shims/registry';
import { getRuntime } from '../_shims/node-runtime';
setShims(getRuntime());

declare module '../_shims/manual-types' {
  export namespace manual {

    export type Agent = types.Agent;

    export import fetch = types.fetch;

    export type Request = types.Request;

    export type RequestInfo = types.RequestInfo;

    export type RequestInit = types.RequestInit;

    export type Response = types.Response;

    export type ResponseInit = types.ResponseInit;

    export type ResponseType = types.ResponseType;

    export type BodyInit = types.BodyInit;

    export type Headers = types.Headers;

    export type HeadersInit = types.HeadersInit;

    export type BlobPropertyBag = types.BlobPropertyBag;

    export type FilePropertyBag = types.FilePropertyBag;

    export type FileFromPathOptions = types.FileFromPathOptions;

    export import FormData = types.FormData;

    export import File = types.File;

    export import Blob = types.Blob;

    export type Readable = types.Readable;

    export type FsReadStream = types.FsReadStream;

    export import ReadableStream = types.ReadableStream;
  }
}
