// mongodb/src/operations/client_bulk_write/executor.ts
import { type Document } from '../../bson';
import { CursorTimeoutContext, CursorTimeoutMode } from '../../cursor/abstract_cursor';
import { ClientBulkWriteCursor } from '../../cursor/client_bulk_write_cursor';
import {
  MongoClientBulkWriteError,
  MongoClientBulkWriteExecutionError,
  MongoInvalidArgumentError,
  MongoServerError
} from '../../error';
import { type MongoClient } from '../../mongo_client';
import { TimeoutContext } from '../../timeout';
import { resolveTimeoutOptions } from '../../utils';
import { WriteConcern } from '../../write_concern';
import { executeOperation } from '../execute_operation';
import { ClientBulkWriteOperation } from './client_bulk_write';
import { ClientBulkWriteCommandBuilder } from './command_builder';
import {
  type AnyClientBulkWriteModel,
  type ClientBulkWriteOptions,
  type ClientBulkWriteResult
} from './common';
import { ClientBulkWriteResultsMerger } from './results_merger';

/**
 * Responsible for executing a client bulk write.
 * @internal
 */
export class ClientBulkWriteExecutor {
  private readonly client: MongoClient;
  private readonly options: ClientBulkWriteOptions;
  private readonly operations: ReadonlyArray<AnyClientBulkWriteModel<Document>>;

  /**
   * Instantiate the executor.
   * @param client - The mongo client.
   * @param operations - The user supplied bulk write models.
   * @param options - The bulk write options.
   */
  constructor(
    client: MongoClient,
    operations: ReadonlyArray<AnyClientBulkWriteModel<Document>>,
    options?: ClientBulkWriteOptions
  ) {
    if (operations.length === 0) {
      throw new MongoClientBulkWriteExecutionError('No client bulk write models were provided.');
    }

    this.client = client;
    this.operations = operations;
    this.options = {
      ordered: true,
      bypassDocumentValidation: false,
      verboseResults: false,
      ...options
    };


    if (!this.options.writeConcern) {
      this.options.writeConcern = WriteConcern.fromOptions(this.client.s.options);
    }

    if (this.options.writeConcern?.w === 0) {
      if (this.options.verboseResults) {
        throw new MongoInvalidArgumentError(
          'Cannot request unacknowledged write concern and verbose results'
        );
      }

      if (this.options.ordered) {
        throw new MongoInvalidArgumentError(
          'Cannot request unacknowledged write concern and ordered writes'
        );
      }
    }
  }

  /**
   * Execute the client bulk write. Will split commands into batches and exhaust the cursors
   * for each, then merge the results into one.
   * @returns The result.
   */
  async execute(): Promise<ClientBulkWriteResult> {


    const pkFactory = this.client.s.options.pkFactory;
    const commandBuilder = new ClientBulkWriteCommandBuilder(
      this.operations,
      this.options,
      pkFactory
    );

    const resolvedOptions = resolveTimeoutOptions(this.client, this.options);
    const context = TimeoutContext.create(resolvedOptions);

    if (this.options.writeConcern?.w === 0) {
      while (commandBuilder.hasNextBatch()) {
        const operation = new ClientBulkWriteOperation(commandBuilder, this.options);
        await executeOperation(this.client, operation, context);
      }
      return ClientBulkWriteResultsMerger.unacknowledged();
    } else {
      const resultsMerger = new ClientBulkWriteResultsMerger(this.options);

      while (commandBuilder.hasNextBatch()) {
        const cursorContext = new CursorTimeoutContext(context, Symbol());
        const options = {
          ...this.options,
          timeoutContext: cursorContext,
          ...(resolvedOptions.timeoutMS != null && { timeoutMode: CursorTimeoutMode.LIFETIME })
        };
        const cursor = new ClientBulkWriteCursor(this.client, commandBuilder, options);
        try {
          await resultsMerger.merge(cursor);
        } catch (error) {




          if (error instanceof MongoServerError && !(error instanceof MongoClientBulkWriteError)) {


            const bulkWriteError = new MongoClientBulkWriteError({
              message: 'Mongo client bulk write encountered an error during execution'
            });
            bulkWriteError.cause = error;
            bulkWriteError.partialResult = resultsMerger.bulkWriteResult;
            throw bulkWriteError;
          } else {

            throw error;
          }
        }
      }


      if (resultsMerger.writeConcernErrors.length > 0 || resultsMerger.writeErrors.size > 0) {
        const error = new MongoClientBulkWriteError({
          message: 'Mongo client bulk write encountered errors during execution.'
        });
        error.writeConcernErrors = resultsMerger.writeConcernErrors;
        error.writeErrors = resultsMerger.writeErrors;
        error.partialResult = resultsMerger.bulkWriteResult;
        throw error;
      }

      return resultsMerger.bulkWriteResult;
    }
  }
}
