// mongodb/src/sdam/server_selection_events.ts
import { HostAddress } from '.././utils';
import {
  SERVER_SELECTION_FAILED,
  SERVER_SELECTION_STARTED,
  SERVER_SELECTION_SUCCEEDED,
  WAITING_FOR_SUITABLE_SERVER
} from '../constants';
import { type ReadPreference } from '../read_preference';
import { type ServerSelector } from './server_selection';
import type { TopologyDescription } from './topology_description';

/**
 * The base export class for all logs published from server selection
 * @internal
 * @category Log Type
 */
export abstract class ServerSelectionEvent {
  /** String representation of the selector being used to select the server.
   *  Defaults to 'custom selector' for application-provided custom selector case.
   */
  selector: string | ReadPreference | ServerSelector;
  
  operation: string;
  
  topologyDescription: TopologyDescription;

  
  abstract name:
    | typeof SERVER_SELECTION_STARTED
    | typeof SERVER_SELECTION_SUCCEEDED
    | typeof SERVER_SELECTION_FAILED
    | typeof WAITING_FOR_SUITABLE_SERVER;

  abstract message: string;

  
  constructor(
    selector: string | ReadPreference | ServerSelector,
    topologyDescription: TopologyDescription,
    operation: string
  ) {
    this.selector = selector;
    this.operation = operation;
    this.topologyDescription = topologyDescription;
  }
}

/**
 * An event published when server selection starts
 * @internal
 * @category Event
 */
export class ServerSelectionStartedEvent extends ServerSelectionEvent {
  
  name = SERVER_SELECTION_STARTED;
  message = 'Server selection started';

  
  constructor(
    selector: string | ReadPreference | ServerSelector,
    topologyDescription: TopologyDescription,
    operation: string
  ) {
    super(selector, topologyDescription, operation);
  }
}

/**
 * An event published when a server selection fails
 * @internal
 * @category Event
 */
export class ServerSelectionFailedEvent extends ServerSelectionEvent {
  
  name = SERVER_SELECTION_FAILED;
  message = 'Server selection failed';
  
  failure: Error;

  
  constructor(
    selector: string | ReadPreference | ServerSelector,
    topologyDescription: TopologyDescription,
    error: Error,
    operation: string
  ) {
    super(selector, topologyDescription, operation);
    this.failure = error;
  }
}

/**
 * An event published when server selection succeeds
 * @internal
 * @category Event
 */
export class ServerSelectionSucceededEvent extends ServerSelectionEvent {
  
  name = SERVER_SELECTION_SUCCEEDED;
  message = 'Server selection succeeded';
  
  serverHost: string;
  
  serverPort: number | undefined;

  
  constructor(
    selector: string | ReadPreference | ServerSelector,
    topologyDescription: TopologyDescription,
    address: string,
    operation: string
  ) {
    super(selector, topologyDescription, operation);
    const { host, port } = HostAddress.fromString(address).toHostPort();
    this.serverHost = host;
    this.serverPort = port;
  }
}

/**
 * An event published when server selection is waiting for a suitable server to become available
 * @internal
 * @category Event
 */
export class WaitingForSuitableServerEvent extends ServerSelectionEvent {
  
  name = WAITING_FOR_SUITABLE_SERVER;
  message = 'Waiting for suitable server to become available';
  
  remainingTimeMS: number;

  
  constructor(
    selector: string | ReadPreference | ServerSelector,
    topologyDescription: TopologyDescription,
    remainingTimeMS: number,
    operation: string
  ) {
    super(selector, topologyDescription, operation);
    this.remainingTimeMS = remainingTimeMS;
  }
}
